File: connector/src/main/java/io/github/dejarol/arcgis/spark/connector/core/http/MismatchingAPIResponseException.java
package io.github.dejarol.arcgis.spark.connector.core.http;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import scala.reflect.Manifest;

/**
 * Indicates that a parsed ArcGIS API response could not be mapped to the expected type.
 *
 * @since 0.1.0
 */
public class MismatchingAPIResponseException
        extends RuntimeException {

    /**
     * Creates an exception describing a mapping failure.
     *
     * @param message description of the mapping failure and received response
     * @param cause   error raised while mapping the response
     * @since 0.1.0
     */
    public MismatchingAPIResponseException(
            String message,
            Throwable cause
    ) {
        super(message, cause);
    }

    /**
     * Creates an exception describing a response that could not be mapped to the expected type.
     *
     * @param apiResponse raw API response that failed to map
     * @param manifest    expected response type
     * @param cause       error raised while mapping the response
     * @param <T>         expected response type
     * @return an exception containing the expected type and formatted response
     * @since 0.1.0
     */
    @Contract("_, _, _ -> new")
    public static <T> @NotNull MismatchingAPIResponseException create(
            @NotNull String apiResponse,
            @NotNull Manifest<T> manifest,
            @NotNull Throwable cause
    ) {

        return new MismatchingAPIResponseException(
            "Received an API response that could not be mapped to " + manifest.toString() + ". " +
            "Response content was\n" + apiResponse,
            cause
        );
    }
}

File: connector/src/main/scala/io/github/dejarol/arcgis/spark/connector/core/http/MismatchingAPIResponseException.scala
package io.github.dejarol.arcgis.spark.connector.core.http

import org.json4s.JValue
import org.json4s.native.JsonMethods.{pretty, render}

/**
 * Indicates that a parsed ArcGIS API response could not be mapped to the expected type.
 *
 * Do not instantiate directly; use [[MismatchingAPIResponseException.create]] instead.
 *
 * @param message description of the mapping failure and received response
 * @param cause   error raised while mapping the response
 * @since 0.1.0
 */
case class MismatchingAPIResponseException private(
                                                    message: String,
                                                    cause: Throwable
                                                  )
  extends RuntimeException(message, cause)

object MismatchingAPIResponseException {

  /**
   * Creates an exception describing a response that could not be mapped to the expected type.
   *
   * @param jValue parsed API response that failed to map
   * @param cause  error raised while mapping the response
   * @tparam A expected response type
   * @return an exception containing the expected type and formatted response
   * @since 0.1.0
   */
  def create[A: Manifest](
                           jValue: JValue,
                           cause: Throwable
                         ): MismatchingAPIResponseException = {

    val valueDescription = implicitly[Manifest[A]].toString()
    MismatchingAPIResponseException(
      f"Received an API response that could not be mapped to $valueDescription. " +
        f"Response content was\n${pretty(render(jValue))}",
      cause
    )
  }
}

